homeassistant/components/enphase_envoy/config_flow.py

```diff
@@ -0,0 +1,166 @@
+"""Config flow for Enphase Envoy integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from envoy_reader.envoy_reader import EnvoyReader
+import httpx
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_IP_ADDRESS,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+ENVOY = "Envoy"
+
+CONF_SERIAL = "serial"
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect."""
+    envoy_reader = EnvoyReader(
+        data[CONF_HOST], data[CONF_USERNAME], data[CONF_PASSWORD], inverters=True
+    )
+
+    try:
+        await envoy_reader.getData()
+    except httpx.HTTPStatusError as err:
+        raise InvalidAuth from err
+    except (AttributeError, httpx.HTTPError) as err:
+        raise CannotConnect from err
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Enphase Envoy."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize an envoy flow."""
+        self.ip_address = None
+        self.name = None
+        self.username = None
+        self.serial = None
+
+    @callback
+    def _async_generate_schema(self):
+        """Generate schema."""
+        schema = {}
+
+        if self.ip_address:
+            schema[vol.Required(CONF_HOST, default=self.ip_address)] = vol.In(
+                [self.ip_address]
+            )
+        else:
+            schema[vol.Required(CONF_HOST)] = str
+
+        schema.update(
+            {
+                vol.Optional(CONF_USERNAME, default=self.username or "envoy"): str,
+                vol.Optional(CONF_PASSWORD, default=""): str,
+            }
+        )
```

Comment:
This is shorter + faster:

```suggestion
        schema[vol.Optional(CONF_USERNAME, default=self.username or "envoy")] = str
        schema[vol.Optional(CONF_PASSWORD, default="")] = str
```